### PR TITLE
MaxMind License Key Support

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
Done
Add MAXMIND_LICENSE_KEY to conf
Display message if MAXMIND_LICENSE_KEY is not set in GUI and cron job

Todo
Testing
Validation of MAXMIND_LICENSE_KEY format
Documentation
Other alerts/log/emails as needed (may not be needed, cron will receive output and will fire off to root)

File: mailscanner/languages/it.php [hunk 1]
```diff
@@ -568,6 +568,7 @@
     'geoipupdate15' => 'GeoIP Database Update',
     'dieproxy15' => 'Proxy type should be either "HTTP" or "SOCKS5", check your configuration file',
     'auditlog15' => 'Ran GeoIP update',
+    'geoipnokey15' => 'A license key from www.maxmind.com is needed to download GeoLite2 data',
```
Comment (endelwar) [inline mailscanner/languages/it.php]: translation: 'È necessaria una chiave di licenza ottenibile da www.maxmind.com per scaricare i dati GeoLite2'

**Comment by ghost** [comment]
Test results with GeoIP Update style key

GUI

```
Downloading file, please wait...
GeoIP data file successfully downloaded
Download complete, unpacking files...

GeoIP data file successfully unpacked
Process completed!
```

Cron

```
Downloading file, please wait...
GeoIP data file successfully downloaded
Download complete, unpacking files...

GeoIP data file successfully unpacked
Process completed!
```

**Comment by ghost** [comment]
Non GeoIP Update license key

```
Downloading file, please wait...
GeoIP data file successfully downloaded
Download complete, unpacking files...

GeoIP data file successfully unpacked
Process completed!
```

Good news is that no matter how a person generates the key from MaxMind, it works either way.

**Comment by ghost** [comment]
Validation added, keys appear to be 16 character alphanumeric key.  We'll start with that.

File: mailscanner/geoip_update.php [hunk 2]
```diff
@@ -58,9 +61,9 @@
     ob_start();
     echo __('downfile15') . '<br>' . "\n";
 
-    $files_base_url = 'http://geolite.maxmind.com';
+    $files_base_url = 'http://download.maxmind.com';
```
Comment (Skywalker-11) [inline mailscanner/geoip_update.php]: Does it also work with https? Might be worth to use it to prevent MitM and resulting attack vectors.

**Comment by oschwald** [comment]
Older style keys are 12 characters. It might be worth supporting those too.

File: mailscanner/geoip_update.php (hunk 2 above)
Comment (ghost) [reply mailscanner/geoip_update.php]: Done, and it works.

File: mailscanner/functions.php [hunk 3]
```diff
@@ -4644,7 +4644,7 @@ function validateInput($input, $type)
                 return true;
             }
         case 'maxmind':
-            if (preg_match('/^[0-9A-Za-z]{16}$/', $input)) {
+            if (preg_match('/^[0-9A-Za-z]{12,16}$/', $input)) {
```
Comment (endelwar) [inline mailscanner/functions.php]: This should be `^([0-9A-Za-z]{12}|[0-9A-Za-z]{16})$` to catch 12 and 16 only
https://regex101.com/r/v4qDCx/1/

Comment (ghost) [reply mailscanner/functions.php]: Done.

**Comment by endelwar** [comment]
Looks good to me, great work @shawniverson!

**Comment by ghost** [comment]
@endelwar PR is out in the wild on the eFa4-Testing repo for the next couple of days.  If any issues arise I will update this PR before giving the all clear to merge.

**Comment by ghost** [comment]
@endelwar All clear. Ready to merge.  Reports coming back positive.  Probably need a blurb to forum and a https://docs.mailwatch.org/ entry once merged and changelog is updated.

**Comment by endelwar** [comment]
Release 1.2.15 published, docs and website updated.

**Comment by atftb** [comment]
So I presume that these changes will be incorporated in a soon to be released download?
I’m a bit confused on what key to generate.  I presume it will be used with geoipupdate which historically wasn’t installed on my mx hosts.  (There is a geoip_update.php script, of course, which was included as a part of MailWatch.)  The maxmind site wants to know which version of geoipupdate to generate a key for, 3.1.1 and newer, or versionsl older than 3.1.1

Or am I barking up the wrong tree?

...Kevin
--
Kevin Miller
Network/email Administrator, CBJ MIS Dept.
155 South Seward Street
Juneau, Alaska 99801
Phone: (907) 586-0242, Fax: (907) 586-4588 Registered Linux User No: 307357

**Comment by Skywalker-11** [comment]
A new release with the updated MaxMind support is available at https://github.com/mailwatch/MailWatch/releases 

When you are logged in on the MaxMind site go to "My License Key". There request a new Key and select that GeoIpUpdate is not used. A licence Key is generated (active after 5-15 min.)

You can check if it works via (replace `xxxxx` with your license key)
```
wget "https://download.maxmind.com/app/geoip_download?edition_id=GeoLite2-Country&suffix=tar.gz&license_key=xxxxx"
```

**Comment by ghost** [comment]
@atftb Look for "MAXMIND_LICENSE_KEY" in conf.php for mailwatch, and set it after obtaining it.  That's it :)  Your geoip_update.php will use it, and if you use the cron job, it will use it too.

**Pull request merged**